onnxruntime/core/providers/js/operators/reduce.cc

```diff
@@ -0,0 +1,95 @@
+      (*KernelDefBuilder::Create())                                                       \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),                     \
+      ReduceOp<float>);                                                                   \
+                                                                                          \
+  ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_EX(                                                \
+      ReduceOp,                                                                           \
+      kOnnxDomain,                                                                        \
+      sinceVersion, endVersion,                                                           \
+      int32_t,                                                                            \
+      kJsExecutionProvider,                                                               \
+      (*KernelDefBuilder::Create())                                                       \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<int32_t>()),                   \
+      ReduceOp<int32_t>);
+
+#define REGISTER_REDUCE_ELEMENTWISE_KERNEL(ReduceOp, sinceVersion)     \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                       \
+      ReduceOp,                                                        \
+      kOnnxDomain,                                                     \
+      sinceVersion,                                                    \
+      float,                                                           \
+      kJsExecutionProvider,                                            \
+      (*KernelDefBuilder::Create())                                    \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<float>())   \
+          .InputMemoryType(OrtMemTypeCPU, 1),                          \
+      ReduceOp<float>);                                                \
+                                                                       \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                       \
+      ReduceOp,                                                        \
+      kOnnxDomain,                                                     \
+      sinceVersion,                                                    \
+      int32_t,                                                         \
+      kJsExecutionProvider,                                            \
+      (*KernelDefBuilder::Create())                                    \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<int32_t>()) \
+          .InputMemoryType(OrtMemTypeCPU, 1),                          \
+      ReduceOp<int32_t>);
+
+REGISTER_REDUCE_ELEMENTWISE_VERSIONSED_KERNEL(ReduceMean, 1, 10);
+REGISTER_REDUCE_ELEMENTWISE_VERSIONSED_KERNEL(ReduceMean, 11, 12);
+REGISTER_REDUCE_ELEMENTWISE_VERSIONSED_KERNEL(ReduceMean, 13, 17);
+REGISTER_REDUCE_ELEMENTWISE_KERNEL(ReduceMean, 18);
```

Comment:
[nit] please add a comment the describe the following information, which may otherwise confuse future maintainers:

macro `REGISTER_REDUCE_ELEMENTWISE_VERSIONSED_KERNEL` does not set `.InputMemoryType(OrtMemTypeCPU, 1)`, so in future if a new opset update applies to Reduce* operators, we may need to add another macro like `REGISTER_REDUCE_ELEMENTWISE_VERSIONSED_KERNEL_WITH_AXIS_IN_INPUT` to set input memory type.

for example, if opset v20 added some change to ReduceMean, then it will become:
```
REGISTER_REDUCE_ELEMENTWISE_VERSIONSED_KERNEL(ReduceMean, 1, 10);
REGISTER_REDUCE_ELEMENTWISE_VERSIONSED_KERNEL(ReduceMean, 11, 12);
REGISTER_REDUCE_ELEMENTWISE_VERSIONSED_KERNEL(ReduceMean, 13, 17);
REGISTER_REDUCE_ELEMENTWISE_VERSIONSED_KERNEL_WITH_AXIS_IN_INPUT(ReduceMean, 18, 19);
REGISTER_REDUCE_ELEMENTWISE_KERNEL(ReduceMean, 20);
```

This is because since opset 18, input[1] exists so it need to be registered differently.